sky/global_user_state.py

```diff
@@ -255,12 +338,64 @@ def set_cluster_metadata(cluster_name: str, metadata: Dict[str, Any]) -> None:
         raise ValueError(f'Cluster {cluster_name} not found.')
 
 
+def get_cluster_history_metadata(cluster_hash: str) -> Optional[Dict[str, Any]]:
+    rows = _DB.cursor.execute(
+        'SELECT metadata FROM cluster_history WHERE cluster_hash=(?)',
+        (cluster_hash,))
+    for (metadata,) in rows:
+        if metadata is None:
+            return None
+        return json.loads(metadata)
+
+
+def set_cluster_history_metadata(cluster_hash: str,
+                                 metadata: Dict[str, Any]) -> None:
+    _DB.cursor.execute(
+        'UPDATE cluster_history SET metadata=(?) WHERE cluster_hash=(?)', (
+            json.dumps(metadata),
+            cluster_hash,
+        ))
+    count = _DB.cursor.rowcount
+    _DB.conn.commit()
+    assert count <= 1, count
+    if count == 0:
+        raise ValueError(f'Cluster hash {cluster_hash} not found.')
+
+
+def get_cluster_hash(cluster_name: str) -> Optional[str]:
+    rows = _DB.cursor.execute(
+        'SELECT cluster_hash FROM clusters WHERE name=(?)', (cluster_name,))
+    for (cluster_hash,) in rows:
+        if cluster_hash is None:
+            return None
+        return cluster_hash
+
+
+def set_cluster_hash(cluster_name: str, cluster_hash: str) -> None:
+    _DB.cursor.execute('UPDATE clusters SET cluster_hash=(?) WHERE name=(?)', (
+        cluster_hash,
+        cluster_name,
+    ))
+    count = _DB.cursor.rowcount
+    _DB.conn.commit()
+    assert count <= 1, count
+    if count == 0:
+        raise ValueError(f'Cluster {cluster_name} not found.')
+
+
+def get_requested_resources(
+        handle: Optional['backends.Backend.ResourceHandle']):
+    launched_resources = handle.launched_resources
+    launched_nodes = handle.launched_nodes
+    return f'{launched_nodes}x {launched_resources}'
```

Comment:
For the `requested_resources`, please refer to the comment above, i.e. should be the resources from the `task` instead of the `handle.launched_resources`.
As mentioned above, we should keep the `launched_reosurces` as well, so as to have more information in the `cluster_history`. However, instead of saving the string here, we may need to save the `pickle.dumps((launched_nodes, launched_resources))`, so that we can recover the `Resources` object from it later. Same for the `requested_resources`.